### PR TITLE
Use app routes

### PR DESCRIPTION
**Pull request by JonathanHallam**
https://trello.com/c/98a0k34y/371-service-discovery-needs-to-use-the-apps-route-rather-than-assume-that-appnamecloudapps-would-work

This allows our prometheus to find apps with non standard or multiple routes. In the event that a route isn't defined it will default to use the app_name as a route. 

If an app that has multiple routes has one deleted whilst it's running it will fall back to the next available route or the app name if none are available.

Only one route is ever exposed to prometheus for scraping. This stops an application generating multiple series of metrics.

### CODE REVIEW

File: lib/cf_app_discovery/service_broker.rb [hunk 1]
```diff
@@ -54,9 +54,13 @@ class ServiceBroker < Sinatra::Application
     delete "/v2/service_instances/:instance_id/service_bindings/:id" do |_, binding_id|
       content_type :json
       service_binding = client.service_binding(binding_id)
-      app_guid = service_binding.fetch(:entity).fetch(:app_guid)
-      cleaner = Cleaner.new(filestore_manager: filestore_manager)
-      cleaner.remove_old_target(app_guid)
+      if service_binding.has_key?(:entity) == false
+        STDERR.puts "service binding: #{service_binding}"
```
Comment (philandstuff) [inline lib/cf_app_discovery/service_broker.rb]: this looks like debug output - can we remove (or make it into a proper error message)

File: lib/cf_app_discovery/client.rb [hunk 2]
```diff
@@ -11,21 +11,42 @@ def apps
...
+      resources.each do |resource|
+        resource[:route] = get_first_route(resource)
+      end
+      resources
     end
 
     def app(app_guid)
-      get("/v2/apps/#{app_guid}")
+      resource = get("/v2/apps/#{app_guid}")
+      resource[:route] = get_first_route(resource)
```
Comment (philandstuff) [inline lib/cf_app_discovery/client.rb]: this pattern (get a resource, then get_first_route on it) seems to be duplicated above. could we extract something common?

File: lib/cf_app_discovery/filestore_manager_factory.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 class CfAppDiscovery
   class FilestoreManagerFactory
     def self.filestore_manager_builder(environment, targets_path)
-      if environment.downcase == 'local'
+      if environment.casecmp('local')
```
Comment (MuriloDalRi) [inline lib/cf_app_discovery/filestore_manager_factory.rb]: I don't see why we need to change this

File: lib/cf_app_discovery/client.rb [hunk 4]
```diff
@@ -11,21 +11,42 @@ def apps
...
+    def routes(service_binding_id)
+      get("/v2/apps/#{service_binding_id}/routes")
+    end
+
   private
 
+    def get_first_route(resource)
+      routes_data = routes(resource.dig(:metadata, :guid)).fetch(:resources)
+      if routes_data.first.nil?
+        resource.dig :entity, :name
```
Comment (philandstuff) [inline lib/cf_app_discovery/client.rb]: what's the reason for using `:name` if there are no routes?  Surely if there are no routes, then the app is simply unreachable, so it's an error condition?

If we're deliberately making a target that will fail, we should at least document that we're doing so.

File: lib/cf_app_discovery/service_broker.rb [hunk 5]
```diff
@@ -54,9 +54,13 @@ class ServiceBroker < Sinatra::Application
     delete "/v2/service_instances/:instance_id/service_bindings/:id" do |_, binding_id|
       content_type :json
       service_binding = client.service_binding(binding_id)
-      app_guid = service_binding.fetch(:entity).fetch(:app_guid)
-      cleaner = Cleaner.new(filestore_manager: filestore_manager)
-      cleaner.remove_old_target(app_guid)
+      if service_binding.has_key?(:entity) == false
```
Comment (philandstuff) [inline lib/cf_app_discovery/service_broker.rb]: can we flip this `if` to put the positive case first?  it's confusing to say `if x == false then foo else bar` rather than `if x then bar else foo`

File: lib/cf_app_discovery/filestore_manager_factory.rb (hunk 3 above)
Comment (kentsanggds) [reply lib/cf_app_discovery/filestore_manager_factory.rb]: The govuk linter was complaining

File: lib/cf_app_discovery/client.rb (hunk 4 above)
Comment (kentsanggds) [reply lib/cf_app_discovery/client.rb]: I had a look into this, when the target updater runs, it will also pick up apps that don't have any routes as it checks against all the apps in PaaS.
Perhaps it would be better to set the route to `nil` or ''?

File: lib/cf_app_discovery/service_broker.rb (hunk 5 above)
Comment (kentsanggds) [reply lib/cf_app_discovery/service_broker.rb]: Sure, will update

File: lib/cf_app_discovery/service_broker.rb (hunk 1 above)
Comment (kentsanggds) [reply lib/cf_app_discovery/service_broker.rb]: Will try to output a proper error message as I was occasionally getting errors on deletion.

File: lib/cf_app_discovery/client.rb (hunk 2 above)
Comment (kentsanggds) [reply lib/cf_app_discovery/client.rb]: sure, will refactor

File: lib/cf_app_discovery/client.rb (hunk 4 above)
Comment (philandstuff) [reply lib/cf_app_discovery/client.rb]: I think there's a legitimate error case where either:

 - a user binds an app that has no routes
 - a user removes the last route from an app that is already bound

I just don't know exactly how we handle this gracefully.  There are options:

 - generate a known-bad target config, so that prometheus shows some failed scrapes
 - don't generate a config (or generate an "inactive" config) - but how does the user know they broke something?

The current behaviour is (by accident rather than design) to generate a config that will fail to scrape and so prometheus will show that the app is down (via the `up` metric).  Maybe that's worth preserving? But if it is, we should document it.

File: spec/cf_app_discovery/target_configuration_spec.rb [hunk 6]
```diff
@@ -71,7 +73,7 @@
         },
       },
       {
-        targets: ["app-1.example.com"],
+        targets: ["route-1.example.com"],
```
Comment (philandstuff) [inline spec/cf_app_discovery/target_configuration_spec.rb]: (argh previous review broke)

This test looks like it hardcodes one single domain (in paas_domain) for all routes. If so, that won't work for custom domains -- it'll only work for cloudapps.digital domains.

Do we need more work to support custom domains? We know that our users need this. (it could be a separate PR if necessary).

File: lib/cf_app_discovery/client.rb (hunk 4 above)
Comment (kentsanggds) [reply lib/cf_app_discovery/client.rb]: ok, lets document the behaviour in the README and make users aware in other docs as well.

File: spec/cf_app_discovery/target_configuration_spec.rb (hunk 6 above)
Comment (philandstuff) [reply spec/cf_app_discovery/target_configuration_spec.rb]: (after offline chat) we've agreed to do custom domains in a separate PR

Comment (JonathanHallam) [reply spec/cf_app_discovery/target_configuration_spec.rb]: 👍

File: README.md [hunk 7]
```diff
@@ -109,3 +109,7 @@ To see the apps bound to a service:
 `cf services`
 
 [user provided service]: https://docs.cloudfoundry.org/devguide/services/user-provided.html#credentials
+
+### Routes
+
+If an app is bound that doesn't have a route or if an app that's already bound has all its routes removed prometheus will not be able to scrape it. This will be recorded in the `up()` metric
```
Comment (philandstuff) [inline README.md]: very nitpicky, but parentheses don't really exist in metric names. I think this would be more accurately listed as `up` rather than `up()`

File: lib/cf_app_discovery/client.rb [hunk 8]
```diff
@@ -17,15 +17,14 @@ def apps
...
+        add_first_route(resource)
       end
       resources
     end
 
     def app(app_guid)
       resource = get("/v2/apps/#{app_guid}")
-      resource[:route] = get_first_route(resource)
-      resource
+      add_first_route(resource)
```
Comment (philandstuff) [inline lib/cf_app_discovery/client.rb]: this isn't quite what I intended... let's chat tomorrow

Comment (philandstuff) [reply lib/cf_app_discovery/client.rb]: we spoke irl and it turns out that what I intended turned out not to be possible
